### ISSUE TITLE
Include validated DFlash restores in retained-prefix admission accounting

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#372]
DFlash can restore a retained target prefix and draft tail, but v0.135.0 always charges the cold full-prompt workspace for this route. This rejects valid warm continuations under memory pressure. Refs #365, #370 and #361. This is admission-accounting work, separate from the tap-storage PR.

Observed on a non-production RTX 5090 after the #370 cold-storage fix:

    cold: 131074 prompt tokens + 1024 output tokens, zero OOM/retries
    retained prefix: 131040 tokens, 4132.5 MB
    warm context reservation: 132200
    warm request cost: 7081 MB plus unchanged reserve
    reclaim parked DFlash session: attainable free rises to 6862 MB
    warm result: HTTP 429 before generation

The exclusion is deliberate qualification scope, not a finding that DFlash cannot benefit. #361 explicitly limited initial qualification to the single-device native MTP restore shape. Release PR #364 says external DFlash retains its existing charges. It is now a known accounting coverage gap, including for co-located workloads on RTX PRO 6000; it is not specific to 32 GB cards.

Exact seams at tag v0.135.0 (001c09e5d):

- `crates/memra-server/src/worker.rs:14411`: `!dspark_drafts.contains_key(&model_key)` excludes DFlash from the retained-prefix admission plan.
- `worker.rs:2689`: `cost_after_prefix_restore` subtracts cold workspace and adds suffix workspace, preserving active context, residual and draft charges; the caller adds unchanged reserve.
- `worker.rs:6540`: the retained-source validator checks the live prefix lease and target geometry, but does not validate a DFlash tail or suffix eligibility.
- `worker.rs:18905`: consuming an admission plan rejects `dspark_prefers_cold`, although that is also the condition used by the DFlash restore route.
- `worker.rs:18926` onward: target KV is copied into a newly allocated full-capacity cache. The retained source is still resident. This is reuse of computed values, not zero-copy sharing and not permission to discount target KV or source residency.
- `worker.rs:19538`: `DflashKv::from_tail` reconstructs draft KV from the retained tail; `worker.rs:19544` consumes the restored target cache in `dspark_spec_session_from_restored`. Success at `worker.rs:19556` reports cached rows and primes only the suffix.
- `worker.rs:19566` and `worker.rs:19574`: failed conversion or inadequate draft tail currently falls through to cold prime. That fallback is unsafe after charging only a suffix.
- `crates/memra-engine/src/dflash.rs:5110`: restored-session construction checks target/draft positions, capacity and boundary logits. `dflash.rs:5243` additionally refuses a nonempty suffix below PRIME_MIN_T; the plan must distinguish full-cover from valid suffix restores.

This is **not a one-line inclusion**. Extend the plan to bind the exact DFlash route, target lease, valid draft tail, boundary logits and supported suffix. Keep source residency, full destination context KV, draft state, conversion transients, residual and reserve charged. Suppress alternate donors after committing the plan. On any planned conversion failure, release both leases and fail closed or re-enter full cold admission before allocating a cold prime. Do not silently force plain.

Required receipt cell: the recorded 128k sampled cold/1024-output request followed by the warm continuation with nonzero cached tokens, DFlash engagement and zero OOM/retries. Also cover full-cover restore, partial/boundary restore, missing/short draft tail, too-short suffix, lost/invalid source and consumed-carrier failure. Fault arms must prove there is no underfunded cold fallback and no leaked lease. Preserve the old-storage tap/draft/logit oracle and greedy restore tape. Validate accounting under a co-located pressure profile before assigning that topology a capacity claim.

Qwen profile: Qwen3.8-27B NVFP4/Q5_K target, q38-dflash2 drafter, 32768-rank FR-Spec trim; MEMRA_CTX=262144, MEMRA_MAX_SESSIONS=4, MEMRA_PREFIX_CACHE_MB=4096, MEMRA_KV_HOST_MB=8192, MEMRA_KV_HOST_VERIFY=1, MEMRA_PRIME_CHUNK=1024, MEMRA_REUSE_POOL_GLOBAL_CAP=1, MEMRA_DSPARK_SPEC=1, MEMRA_DSPARK_PREFIX_RESTORE=1, MEMRA_DSPARK_BOUNDARY_CAPTURE=1, MEMRA_DSPARK_PARTIAL_RESTORE=1, MEMRA_PREFIX_STABLE_BOUNDARY=1. Full vision tower loaded; text requests supply no sampling fields and cap output at 1024. No reserve or scope reduction.

Measured binary SHA-256: a972c62204b0a75ecd5ff884808fdc4c37b50f73aae87790ffb2ebadccdc5df5. Public summary: `research/dflash-tap-storage-20260908/qualification.json` in #370. Full private receipt: Darklanes `research/qwen-ornith-5090-capacity-20260908/qwen-128k-decode-capacity/`, with analysis in `issue365/REPORT.md`.

**Comment by avifenesh** [comment]
CLAIMED 5090-admission 2026-09-08T20:12:43Z branch fix/dflash-retained-admission-20260908

**Comment by avifenesh** [comment]
Banked as draft #377 at b3c411d24a25e36ffdf83911ddab567beb82266a, not merged. Restore validation, seven HTTP fault cells, oracle/tape exactness, pressure readback and software gates passed. The required 128k cold/1024-output plus warm cell remains blocked: three candidate cold trials OOMed before warm admission. The clean main control completed cold and returned 429 warm; omitted seeds differ, so the cold failure still needs investigation. Receipts and the self-review are in the PR. GPU empty, lock free, all owned server PIDs absent and remote scratch removed. Releasing under-work after banking; #372 stays open. No release or fleet actions.

**Comment by avifenesh** [comment]
CLAIMED 5090-admission 2026-09-08T21:06:08Z branch fix/dflash-retained-admission-20260908. One remaining sequence check: replay only the recorded cold and warm requests, without the copied harness's extra calibration request. Main and candidate use identical profiles. Draft stays unmerged pending that result.

**Comment by avifenesh** [comment]
merged in 6285210078609c7e11aa23ae070ad581c184c153; ships in the next batched release; fleet exposure deferred by owner

**Comment by avifenesh** [comment]
State of this issue on today's `main` (`fa5185d90`), read from the tree and re-tested on the CPU (WP-B day 21, branch `lane/spill-b-20260919`).

**The ask is landed.** #377 (`628521007`, 2026-09-08, in v0.137.0) replaced the `!dspark_drafts.contains_key(&model_key)` exclusion with a typed plan: `AdmissionRestoreRoute::{Native, Dflash { ctx_cap, suffix }}` where `DflashRestoreSuffix::new(rows, prompt, partial)` is `FullCover` at an exact cover, `Prime(n)` only when the partial door is on and `n >= PRIME_MIN_T`, and `None` otherwise; `PrefixCache::admission_dflash_restore` binds the live lease, the exact token boundary, a validated draft tail (`tail.len == rows`, `validate_restore`), finite boundary logits of the model's vocab and `dspark_spec_prompt_fits`; the plan is revalidated at consumption (`worker.rs`, the `if let Some(plan) = admission_restore` block) and refused with a 429 if the route, the paid `ctx_cap` or the rows moved. The cost rule is the native restore's `cost_after_prefix_restore` (cold prefill workspace out, suffix workspace in; context, residual, draft state and reserve unchanged), with one difference for DFlash: `workspace_rows = rows.min(prompt_len - PRIME_MIN_T)`, so a full cover keeps a minimum-prime workspace for conversion and boundary sampling. `finish_planned_dflash_conversion` releases the serving lease and returns a rate-limit error when a planned conversion fails, so a suffix-only charge can never fall through to a cold prime. No new `MEMRA_*` read; no new numeric program (the restore reuses computed values, the target KV is copied into a full-capacity cache).

**CPU gate on the arithmetic, run today on the merged tree** (`cargo test --release -p memra-server --lib -- dflash_retained cost_after_prefix_restore`, receipt `research/spill-b-20260919/rtx5090-day21/cpu-tests-372/`):

```text
test worker::tests::dflash_retained_consumed_carrier_failure_releases_serving_lease_and_refuses_cold ... ok
test worker::tests::dflash_retained_suffix_distinguishes_full_boundary_and_unsupported_suffix ... ok
test worker::tests::dflash_retained_plan_checks_route_before_discounting ... ok
test result: ok. 3 passed; 0 failed; 1 ignored; 0 measured; 770 filtered out
```

(the ignored one is `dflash_retained_gpu_plan_fault_matrix`, which needs an exclusively locked CUDA device and the DFlash restore profile.)

**Serving-shape receipt.** The required cell (the recorded 128k sampled cold/1024-output request followed by the warm continuation) is in #377's record: `research/dflash-retained-admission-20260908/final-qualification.json`, warm `cached_tokens=131040`, `restored=131040 suffix=72 cost 7250MB -> 4129MB; reserve 1720MB unchanged; cold fallback forbidden`, DFlash engaged, zero OOM/retries, against the base control's `HTTP 429`. Not re-run today: reproducing the 429 needs a pre-#377 binary, and the bounded twin of that cell on this rig (`MEMRA_DSPARK_SPEC=1 MEMRA_DSPARK_DRAFT=<q38-dflash2 export> MEMRA_DSPARK_PREFIX_RESTORE=1`, a cold prompt then its continuation at a size that puts a 24 GB card under admission pressure, `[admission] retained prefix plan` on the warm turn, restored render equal to the cold render) is pre-registered in `research/spill-b-20260919/DAY21.md` and was not run inside today's budget. The DFlash artifact and its draft exist on this rig, so the cell is runnable when the lead schedules it.

Nothing here changes code; the issue can close on #377's record unless the lead wants the bounded rerun first.

**Comment by avifenesh** [comment]
Landed by #377 (`628521007`, v0.137.0): `AdmissionRestoreRoute::Dflash`, `DflashRestoreSuffix`, `admission_dflash_restore`, revalidation at consumption, `cost_after_prefix_restore` with the min-prime workspace rule; the cited `!dspark_drafts.contains_key` exclusion is gone. Verified by lane B day 21: the three CPU plan tests pass on today's tree (receipt `research/spill-b-20260919/rtx5090-day21/cpu-tests-372/`); the bounded serving-shape cell is pre-registered in its DAY21.md and not run (a 429 reproduction needs a pre-#377 binary). Closing on #377's record; integrated as #612 (merge 5804cac6a).
